src/native/libs/System.Globalization.Native/pal_timeZoneInfo.m

```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+#include "pal_errors_internal.h"
+#include "pal_locale_internal.h"
+#include "pal_timeZoneInfo.h"
+#import <Foundation/Foundation.h>
+
+#if !__has_feature(objc_arc)
+#error This file relies on ARC for memory management, but ARC is not enabled.
+#endif
+
+#if defined(TARGET_MACCATALYST) || defined(TARGET_IOS) || defined(TARGET_TVOS)
+
+/*
+Gets the localized display name that is currently in effect for the specified time zone.
+*/
+int32_t GlobalizationNative_GetTimeZoneDisplayNameNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* timeZoneId, int32_t timeZoneIdLength,
+                                                         TimeZoneDisplayNameType type, uint16_t* result, int32_t resultLength)
+{
+    @autoreleasepool
+    {
+        NSString* tzName = [NSString stringWithCharacters: timeZoneId length: timeZoneIdLength];
+        NSTimeZone* timeZone = [NSTimeZone timeZoneWithName:tzName];
+        NSString* timeZoneName;
+
+        if (type == TimeZoneDisplayName_TimeZoneName)
+            timeZoneName = timeZone.name;
+        else
+        {
+            NSLocale *currentLocale;
+            if(localeName == NULL || lNameLength == 0)
+            {
+                currentLocale = [NSLocale systemLocale];
+            }
+            else
+            {
+                NSString *locName = [NSString stringWithCharacters: localeName length: lNameLength];
+                currentLocale = [NSLocale localeWithLocaleIdentifier:locName];
+            }
+            NSTimeZoneNameStyle style;
+
+            switch (type)
+            {
+                case TimeZoneDisplayName_Standard:
+                    style = NSTimeZoneNameStyleStandard;
+                    break;
+                case TimeZoneDisplayName_DaylightSavings:
+                    style = NSTimeZoneNameStyleDaylightSaving;
+                    break;
+                case TimeZoneDisplayName_Generic:
+                    style = NSTimeZoneNameStyleGeneric;
+                    break;
+                default:
+                    return UnknownError;
+            }
+
+            timeZoneName = [timeZone localizedName:style locale:currentLocale];
+        }
+        if (timeZoneName == NULL || timeZoneName.length == 0)
+        {
+            return UnknownError;
+        }
+
+        int32_t index = 0, dstIdx = 0, resultCode = Success;
+        uint16_t dstCodepoint;
+        while (index < timeZoneName.length)
+        {
+            dstCodepoint = [timeZoneName characterAtIndex: index];
+            Append(result, dstIdx, resultLength, dstCodepoint, resultCode);
```

Comment:
If we are using `Append`, should we include `pal_icushim_internal.h` at the top instead of relying on `pal_local_internal.h` to include it?

Do we want to add a check for `resultCode` to return immediately when we hit an error?

Reply:
Thanks, updated it.